Добавлена функциональность по ТЗ

discomfort-pr opened the pull request:
-Добавлен метод remove в интерфейсе HistoryManager и реализован в классе InMemoryHistoryManager 
-В классе InMemoryHistoryManager добавлены методы linkLast() и removeNode(), исправлены методы add() и getHistory() 
-В классе InMemoryTaskManager скорректированы методы 
-Поле epicId в классе Subtask теперь final
-Добавлены тесты

discomfort-pr committed:
-Добавлен конструктор с подзадачами в классе Epic
-Добавлен метод remove в интерфейсе HistoryManager и реализован в классе InMemoryHistoryManager
-В классе InMemoryHistoryManager добавлены методы linkLast() и removeNode(), исправлены методы add() и getHistory()
-В классе InMemoryTaskManager скорректированы методы
-Поле epicId в классе Subtask теперь final
-Добавлены тесты

slavrikk commented on src/managers/InMemoryHistoryManager.java:
```diff
@@ -1,25 +1,84 @@
 
 public class InMemoryHistoryManager implements HistoryManager {
-    private final List<Integer> history;
+    class Node {
```
1) Внутренний класс лучше вынести в конец класса `InMemoryHistoryManager`
2) Лучше указать `private static`

slavrikk commented on src/managers/InMemoryHistoryManager.java:
```diff
@@ -1,25 +1,84 @@
+    class Node {
+        private Node next;
+        private Node prev;
+        private final int taskId;
```
Тут лучше хранить `Task`

slavrikk commented on src/managers/InMemoryHistoryManager.java:
```diff
@@ -1,25 +1,84 @@
     }
 
     @Override
     public void add(int taskId) {
```
Лучше добавлять непосредственно таску

slavrikk commented on src/managers/InMemoryHistoryManager.java:
```diff
@@ -1,25 +1,84 @@
     }
 
     @Override
     public List<Integer> getHistory() {
```
Тут следует возвращать массив тасок

discomfort-pr committed:
-класс Node теперь private static
-узлы хранят таску вместо айди
-метод add() в InMemoryHistoryManager добавляет саму таску
-метод getHistory() в InMemoryHistoryManager возвращает массив тасок
-тесты и интерфейсы адаптированы под изменения

discomfort-pr commented:
подправил по комментариям

slavrikk commented on src/managers/HistoryManager.java:
```diff
@@ -1,9 +1,11 @@
-    List<Integer> getHistory();
+    void remove(int taskId);
+
+    Task[] getHistory();
```
Лучше возвращать лист `List<Task>`

slavrikk commented on src/managers/InMemoryHistoryManager.java:
```diff
@@ -1,25 +1,86 @@
     @Override
-    public List<Integer> getHistory() {
+    public Task[] getHistory() {
+        Task[] history = new Task[nodes.size()];
```
Лучше возвращать лист `List<Task>`

slavrikk commented on src/managers/InMemoryHistoryManager.java:
```diff
@@ -1,25 +1,86 @@
+    }
+
+    private void removeNode(Node removed) {
+        if (removed == head) {
```
Объекты следует сравнивать через `equals()`

slavrikk commented on src/managers/InMemoryTaskManager.java:
```diff
@@ -176,11 +180,21 @@ private void updateEpicStatus(int epicId) {
 
-    public List<Integer> getHistory() {
-        return historyManager.getHistory();
+    private boolean historyContains(Task target) {
```
Кажется эту проверку лучше вынести в метод `historyManager.remove()`

discomfort-pr committed:
-методы getHistory() интерфейсов возвращают List<Task>
-проверка наличия таски в истории перенесена в метод remove() класса InMemoryHistoryManager
-проверка узлов на равенство в методе removeNode() класса InMemoryHistoryManager теперь через equals()

discomfort-pr commented:
исправил возвращаемый тип, проверку на равенство и вынес проверку на наличие таски в метод remove() класса InMemoryHistoryManager
